### DISCUSSION TITLE
movie play buttons

### DISCUSSION THREAD

**Discussion by adedner** [pyvista/pyvista#4525] [Q&A] [answered]
Hi.
I just came across this project and have been playing around with it - so far it looks great and the usage is intuitive as advertised, thanks for the work.

I work on the open source finite-element software Dune and one thing we often have is time series data let's say on a fixed grid. I am looking for ways of visualizing this. I figured out how to use a slider widget (a `string`  version to get discrete points in time) but what I am looking for would be a _movie control_ widget similar to what paraview has, i.e., _play_, _forward/backward one frame_, and a field to enter a frame number.

Does something like that exist or is there some way I could add this?
Thanks again!

PS: I just figured out that what I am looking for is support for the `vtkPlaybackWidget` and then I found a previous similar question
https://github.com/pyvista/pyvista-support/issues/317
Sorry for missing that. There is probably no new development with this?

**Comment by tkoyama010** [comment]
No, this extension has not been developed yet. See #4402.
The following comments using ipywidgets may be of interest to you.
https://github.com/pyvista/pyvista/issues/4402#issuecomment-1556240732

**Reply by adedner** [reply]
Thanks for the quick reply, much appreciated! After finding #317 I started playing around with `ipywidgets` and together with #4402 I got something that works quite well. Now I'll look into `qtwidgets` to  see how to add similar buttons to the `BackgroundPlotter`. 

One issue I came across when looking at visualizing data sequences:
if I am plotting a slice for example, do I need to recompute the whole slice grid in each step of the sequence after updating the scalar? So if I have
```
plot.add_mesh( mesh.slice_orthogonal(), name="slice" )
```
and then when changing my time slider I do
```
mesh.point_data["value"] = ...
```
do I then need to also add
```
plot.add_mesh( mesh.slice_orthogonal(), name="slice" )
```
or is there a way to `update` the data in the slice without recomputing the slice grid? I'm assuming the connection between the slice and the original grid is not available anymore but I just wanted to check.

**Reply by banesullivan** [reply]
This is best handle through custom algorithms. At this time, we don't expose the `slice_orthogonal()` filter as an algorithm so your current approach is going to be the simplest implementation

**Comment by adedner** [comment]
I hope it's okay if I continue asking slightly off topic questions (on the topic of pyvista but not on the topic in the question title).
If you'd prefer I can close this and open another question.

I have a plotter setup with sliders and plane widgets etc. and would like to screenshot this but without the widgets. I was hoping for some `plotter.disable_widgets` method but that does not seem to exist. I would like to keep the widgets for later use but remove them just for saving an image (or movie - to keep it on topic :-)

**Reply by banesullivan** [reply]
You bring up an excellent request! Removing the widgets has proven tough in the past so the API to remove them may/may not work... but they can be seen in this `close()` method (these are not documented well)

https://github.com/pyvista/pyvista/blob/0e8df552f753362db8fed9dee93e7a7b1dec7f17/pyvista/plotting/widgets.py#L2344-L2351

I have another thought on how this ould be achieved: I think we currently use rendering layers in PyVista and perhaps we could switch the widgets to render in a specific layer and have the `screenshot()` method be able to only take a screenshot of a specific layer, thus not capturing the widgets.

**Reply by adedner** [reply]
I tried the `clear` methon (I love `dir` in python :-) but that seemed to crash the kernel - I didn't investigate fully.
The `layer` idea sounds great. I also just found something else that works using the `On/Off` methods on the widgets.
That could be easily done I guess for all widgets registered with the plotter in some `disable` method:
Perhaps even better to add a parameter to the `plotter.add_w*` functions like `add2screenshot=True`.

Here is an example:

```python
import pyvista as pv
pv.set_plot_theme("document")

from pyvista import examples
vol = examples.download_brain()

plotter = pv.Plotter()

# this is needed to get a good 'origin' for the plane widget
plotter.add_mesh(vol.outline_corners(), color='k') 

# add a slice with plane widget
def slice(normal,origin):
    slices = vol.slice(normal=normal, origin=origin)
    plotter.add_mesh(slices, name="slice")    
plane = plotter.add_plane_widget(slice)

# use second widget to have the clip plane visible in screenshot without the widgets
# this is just to see if that is possible - the white 'outline'  is a bit surprising since it is larger than the domain
plane2 = plotter.add_plane_widget(lambda *args:[], origin_translation=False, normal_rotation=False, implicit=True)
# turn off for now
plane2.Off()

plotter.show()

# now set a specific normal (and if required origin) and make a screenshot without the widget
plane.SetNormal([-1,1,1])
slice(plane.GetNormal(),plane.GetOrigin())
plane.Off()
plane2.On()
plane2.SetNormal( plane.GetNormal() )
plane2.SetOrigin( plane.GetOrigin() )
plotter.render()
plotter.screenshot("test.png", return_img=False)
plane2.Off()
plane.On()
plotter.render()
```
![test](https://github.com/pyvista/pyvista/assets/21035948/006cd0e2-1f71-42ea-b844-5661ebfe444e)

Perhaps now is a good time to abandon this thread and open a feature request.

**Reply by banesullivan** [reply]
Aha, the On/Off switches are probably the easiest way to do this. Nice work!

**Comment by adedner** [comment, marked as answer]
Here is what I ended up adding using `ipywidgets`. I've not used `ipwidgets` before so I'm sure this can be improved but it might serve as starting point for others coming across this issue:

```
import ipywidgets as widgets
from traitlets import CInt

def playBackWidget(minValue,maxValue,callBack):
    class Counter(widgets.DOMWidget):
        value = CInt(minValue, sync=True)
    counter = Counter()

    def button_plus(name):
        if counter.value<maxValue:
            counter.value += 1
    wplus = widgets.Button(icon="plus", layout=widgets.Layout(width='auto'), style={"button_width": "auto"})
    wplus.on_click(button_plus)

    def button_minus(name):
        if counter.value>minValue:
            counter.value -= 1
    wminus = widgets.Button(icon="minus", layout=widgets.Layout(width='auto'), style={"button_width": "auto"})
    wminus.on_click(button_minus)
    
    slider = widgets.IntSlider(value=minValue,min=minValue,max=maxValue,step=1,
                               description="time step",
                               continuous_update=False,
                               readout=False)
    play = widgets.Play(value=minValue, min=minValue, max=maxValue, step=1,
                        interval=100, disabled=False)
    inBox = widgets.BoundedIntText(value=minValue, min=minValue, max=maxValue,
                                   step=1, disabled=False,
                                   layout=widgets.Layout(width='100px'))

    out = widgets.Output(layout={'border': '1px solid black'})    
    play.observe(lambda change: callBack(out,change.new), "value")
    widgets.link((slider, 'value'), (counter, 'value'))
    widgets.link((slider, 'value'), (play, 'value'))
    widgets.link((slider, 'value'), (inBox, 'value'))
    return widgets.VBox(
             [widgets.HBox([wminus, play, wplus, slider, inBox]), out]
           )
````
